src/sidebar/services/oauth-auth.js

```diff
@@ -167,9 +167,27 @@ function auth($http, $rootScope, $window, OAuthClient,
     if (!tokenInfoPromise) {
       var cfg = serviceConfig(settings);
 
+      if (cfg && cfg.grantToken === 'postMessage') {
+        tokenInfoPromise = new Promise(function(resolve, reject) {
+          /** Receive postMessage() response with grant token from LMS app. */
+          function receiveMessage(event) {
+            if (event.origin !== "http://localhost:8001") {
+              return;
+            }
+
+            oauthClient().then(function(client) {
+              client.exchangeGrantToken(event.data).then(function(accessToken) {
+                resolve(accessToken)
+              });
+            });
+          }
+          window.addEventListener("message", receiveMessage, false);
+          window.parent.parent.postMessage("gimme_grant_token", "http://localhost:8001");
```

Comment:
I haven't implemented any kind of timeout, if the response postMessage doesn't come. We should add a very generous timeout, and have the client show an error message and report an error to Sentry if it times out.

I _don't_ think we want to fall back on other methods of login or continue booting up if the postMessage method times out, we want to fail the client startup entirely:

* postMessage login should only ever be enabled in the LMS context (and any future contexts that use it)
* When in the LMS context postMessage login should _always_ be enabled and should always succeed. It would only fail if something has gone badly wrong
* When in the LMS context we would never want to use any other type of login, e.g. login to a first-party account, we only and always want to use the LMS login integration

Reply:
Oh, and this relies on the fact that in the LMS context `window.parent.parent` is the LMS app's iframe, which is of course brittle and not generalised. This should be replaced by iterating over all the parents and/or all the frames reachable and messaging each one of them, until one responds. [The old LTI app's `postMessage()` feature did this](https://github.com/hypothesis/lti/blob/1f8896ee045018e2ac57945f9f1d396f8ab28325/lti/templates/includes/submission_form.html.jinja2#L23-L37) in the opposite direction (the LTI app finding the client's iframe, rather than vice-versa).